src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -58,26 +62,71 @@ def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
     return QueryGenerator(plan.session, snowflake_create_table_plan_info)
 
 
-def replace_child(
+def resolve_and_update_snowflake_plan(
+    node: SnowflakePlan, query_generator: QueryGenerator
+) -> None:
+    """
+    Re-resolve the current snowflake plan if it has a source plan attached, and update the fields with
+    newly resolved value.
+    """
+
+    if node.source_plan is None:
+        return
+
+    new_snowflake_plan = query_generator.resolve(node.source_plan)
+
+    # copy over the newly resolved fields to make it an in-place update
+    node.queries = new_snowflake_plan.queries
+    node.post_actions = new_snowflake_plan.post_actions
+    node.expr_to_alias = new_snowflake_plan.expr_to_alias
+    node.is_ddl_on_temp_object = new_snowflake_plan.is_ddl_on_temp_object
+    node._output_dict = new_snowflake_plan._output_dict
+    node.df_aliased_col_name_to_real_col_name = (
+        new_snowflake_plan.df_aliased_col_name_to_real_col_name
+    )
+    node.placeholder_query = new_snowflake_plan.placeholder_query
+    node.referred_ctes = new_snowflake_plan.referred_ctes
+    node._cumulative_node_complexity = new_snowflake_plan._cumulative_node_complexity
+
+
+def resolve_node(node: LogicalPlan, query_generator: QueryGenerator) -> SnowflakePlan:
+    resolved_node = query_generator.resolve(node)
+    resolved_node._is_valid_for_replacement = True
+
+    return resolved_node
+
+
+def replace_child_and_update_parent(
     parent: LogicalPlan,
```

Comment:
I think we should keep replace_child and update parent snowflake plan separate. The reason that they are together now is because we don't want to keep the snowflake plan to have inconsistent queries. But in this function we are not completely addressing the inconsistency problem. Once parent is updated with correct queries, we need to also update the parent's parent.

Since that will be done separately anyways, what's the advantage of combining the two functions here?

Reply:
this is trying to make the node consistent from parent to child, the benefit of making them together is that whenever child is updated, the corresponding parent node needs to be updated, this is to avoid we forgot to update anything. 
Do you have a use case that you want to handle them separately? maybe we can refactor when you need this, which can make things little bit cleaner about why you need this?

Reply:
in repeated subquery elimination you update ancestors using `_update_parents` in a bottom-up traversal. When a with_query block is added we call `self._update_parents(node, resolved_with_block)` to make the child replacement and updating the queries and mark the parents to be updated later. Then for the updated parents, we call `self._update_parents(node, node)`. 

I think it would be cleaner if we do something like
```
            if node in self._duplicated_nodes:
                # create a WithQueryBlock node
                with_block = WithQueryBlock(
                    name=random_name_for_temp_object(TempObjectType.CTE), child=node
                )
                with_block._is_valid_for_replacement = True

                resolved_with_block = resolve_node(with_block, self._query_generator)
                replace_child(node, resolved_with_block)
                self._update_parents(node)
            elif node in self._updated_nodes:
                # if the node is updated, make sure all nodes up to parent is updated
                self._update_parents(node)
```

in large query breakdown, I am following a similar pattern. We can still use `replace_child_and_update_parent(node, node)` we aren't really replacing a child. Only using it to re-resolve parent queries.

Reply:
agree

Reply:
i actually updated it to two function